src/cargo/util/toml.rs

```diff
@@ -760,19 +766,31 @@ fn normalize(lib: &Option<TomlLibTarget>,
               .set_for_host(toml.plugin.unwrap_or(t2.for_host()));
     }
 
-    fn lib_target(dst: &mut Vec<Target>, l: &TomlLibTarget,
-                  metadata: &Metadata) {
-        let path = l.path.clone().unwrap_or_else(|| {
+    fn lib_target(dst: &mut Vec<Target>,
+                  l: &TomlLibTarget,
+                  metadata: &Metadata,
+                  warnings: &mut Vec<String>) {
+        let path = l.path.clone().unwrap_or(
             PathValue::Path(Path::new("src").join(&format!("{}.rs", l.name())))
-        });
-        let crate_types = l.crate_type.clone().and_then(|kinds| {
-            kinds.iter().map(|s| LibKind::from_str(s))
-                 .collect::<CargoResult<_>>().ok()
-        }).unwrap_or_else(|| {
-            vec![if l.plugin == Some(true) {LibKind::Dylib} else {LibKind::Lib}]
-        });
-
-        let mut target = Target::lib_target(&l.name(), crate_types.clone(),
+        );
+        let crate_types = match l.crate_type.clone() {
+            Some(kinds) =>
+                // For now, merely warn about invalid crate types.
+                // In the future, it might be nice to make them errors.
+                kinds.iter().filter_map(|s| {
+                        let kind = LibKind::from_str(s);
+                        if let Err(ref error) = kind {
+                            warnings.push(format!("warning: {}", error))
+                        }
+                        kind.ok()
+                    }).collect(),
+            None => vec![
+                    if l.plugin == Some(true) {LibKind::Dylib}
+                    else {LibKind::Lib}
+                ]
+        };
```

Comment:
Oh I understand what you meant now.  Done.  Needed a re-test anyway because Travis is having trouble.